### PR TITLE
Basic notebook model + API

### PR DESCRIPTION
**Pull request by wlach**

### CODE REVIEW

**Comment by wlach** [comment]
Had a look at the current prototype with @jaredkerim earlier today. For preliminary feedback, he recommended using slugs (vs. ids) to identify notebooks. Similarly, we'd do better to identify notebook content via a hash of their content.

File: server/notebooks/tests/test_notebook_api.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    assert resp.json() == [{
+        'created': '2018-07-01T13:00:00Z',
+        'id': revision.id,
+        'title': revision.title
+    } for revision in NotebookRevision.objects.filter(notebook=test_notebook)]
+
+    # verify that we can get the details of a single revision as well
+    test_revision = NotebookRevision.objects.filter(notebook=test_notebook).first()
+    resp = client.get(reverse('notebook-revisions-detail', kwargs={'notebook_id': test_notebook.id,
+                                                                   'pk': test_revision.id}))
```
Comment (jaredkerim) [inline server/notebooks/tests/test_notebook_api.py]: Too many spaces make baby guido cry 😢

Comment (jaredkerim) [reply server/notebooks/tests/test_notebook_api.py]: I've been using this lately it works great:

https://github.com/ambv/black

File: server/notebooks/views.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    latest_revision = NotebookRevision.objects.filter(
+        notebook=notebook).order_by('-created').first()
+    if request.user.is_authenticated:
+        user_info = json.dumps(get_user_info_dict(request.user))
+    else:
+        user_info = json.dumps({})
+    return HttpResponse(
+        template.render({
+            'user_info': user_info,
+            'jsmd': latest_revision.content
```
Comment (jaredkerim) [inline server/notebooks/views.py]: We need to make sure that when a notebook is created an initial revision is created so this doesn't fail

File: server/notebooks/views.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+from server.views import get_user_info_dict
+from .models import (Notebook, NotebookRevision)
+
+
+def notebook_view(request, pk):
+    template = loader.get_template('notebook.html')
+    notebook = get_object_or_404(Notebook, pk=pk)
+    latest_revision = NotebookRevision.objects.filter(
+        notebook=notebook).order_by('-created').first()
```
Comment (jaredkerim) [inline server/notebooks/views.py]: You could move .order_by('-created') to 

```py
class Meta:
    ordering = ('-created',)
```

on the model if you always want to order by most recent revision first

File: server/notebooks/models.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+
+class NotebookRevision(models.Model):
+    '''
+    A revision of a specific notebook
+    '''
+
+    id = models.BigAutoField(primary_key=True)
+
+    notebook = models.ForeignKey(Notebook, on_delete=models.CASCADE)
```
Comment (jaredkerim) [inline server/notebooks/models.py]: Add 

```py
related_name='revisions'
```

File: server/notebooks/views.py (hunk 3 above)
Comment (jaredkerim) [reply server/notebooks/views.py]: Then, with the related_name, this would become

```py
notebook.revisions.first()
```

File: server/urls.py [hunk 5]
```diff
@@ -10,6 +10,9 @@
 urlpatterns = [
     url(r'^$', server.views.index, name='index'),
 
+    url(r'^notebooks/', include('server.notebooks.urls')),
+    url(r'^api/', include('server.notebooks.api_urls')),
```
Comment (jaredkerim) [inline server/urls.py]: definitely add a v1 to this, it'll save hassle in the future

File: server/urls.py [hunk 6]
```diff
@@ -10,6 +10,9 @@
 urlpatterns = [
     url(r'^$', server.views.index, name='index'),
```
Comment (jaredkerim) [inline server/urls.py]: Move this to the bottom generally

File: server/views.py [hunk 7]
```diff
@@ -7,7 +7,7 @@
 from social_django.models import UserSocialAuth
 
 
-def _get_user_info_dict(user):
+def get_user_info_dict(user):
     user_social_auth = UserSocialAuth.objects.get(user=user)
     social_auth_extra_data = user_social_auth.extra_data
```
Comment (jaredkerim) [inline server/views.py]: You could also just do this in the middleware so you don't look this up twice

File: server/notebooks/api_views.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        fields = ('id', 'owner', 'title')
+
+
+class NotebookDetailSerializer(NotebookSerializer):
+
+    latest_revision = NotebookLatestRevisionField(read_only=True)
+
+    class Meta:
+        model = Notebook
+        fields = ('id', 'owner', 'title', 'latest_revision')
```
[the remark this answers is not in the document]
Comment (wlach) [reply server/notebooks/api_views.py]: This doesn't work, it complains that there is no property called `many`. It seems like using multiple serializers for different views is an accepted solution in rest framework land, and seems a little easier to understand. I'm inclined to go with that (with the naming tweaks we discussed) for now.

https://stackoverflow.com/a/44070743/295132

File: server/notebooks/api_views.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+
+class NotebookRevisionSerializer(serializers.ModelSerializer):
+    '''
+    Summarization of revisions for a notebook (does not include content)
+    '''
+
+    def create(self, validated_data):
+        return NotebookRevision.objects.create(**{**validated_data, **self.context})
```
[the remark this answers is not in the document]
Comment (wlach) [reply server/notebooks/api_views.py]: yep, leftover from some earlier work

File: server/notebooks/api_views.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        return NotebookRevision.objects.filter(
+            notebook=self.kwargs['notebook_id'])
+
+    def get_serializer_class(self):
+        if self.action == 'list':
+            return NotebookRevisionSerializer
+        return NotebookRevisionDetailSerializer
+
+    def perform_create(self, serializer):
+        serializer.save(**self.get_serializer_context())
```
[the remark this answers is not in the document]
Comment (wlach) [reply server/notebooks/api_views.py]: Yup, I needed to do that. Added a unit test validating this as well.

File: server/notebooks/tests/test_notebook_api.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+def test_notebook_list(client, two_test_notebooks):
+    resp = client.get(reverse('notebooks-list'))
+    assert resp.status_code == 200
+    assert resp.json() == [
+        {
+            "id": notebook.id,
+            "owner": "testuser1",
+            "title": notebook.title
+        } for notebook in Notebook.objects.all()
```
[the remark this answers is not in the document]
Comment (wlach) [reply server/notebooks/tests/test_notebook_api.py]: done

**Pull request merged**